atlasaction/action.go

```diff
@@ -383,6 +385,133 @@ func (a *Actions) SchemaTest(ctx context.Context) error {
+func (a *Actions) SchemaPlan(ctx context.Context) error {
+	err := a.RequiredInputs("config", "env", "repo", "from")
+	if err != nil {
+		return err
+	}
+	tc, err := a.GetTriggerContext()
+	if err != nil {
+		return fmt.Errorf("unable to get the trigger context: %w", err)
+	}
+	params := &atlasexec.SchemaPlanParams{
+		ConfigURL: a.GetInput("config"),
+		Env:       a.GetInput("env"),
+		Vars:      a.GetVarsInput("vars"),
+		Context:   a.GetRunContext(ctx, tc),
+		DevURL:    a.GetInput("dev-url"),
+		From:      a.GetArrayInput("from"),
+		Repo:      a.GetInput("repo"),
+	}
+	params.DryRun = true // Dry-run to get the desired hash
+	plan, err := a.Atlas.SchemaPlan(ctx, params)
+	if err != nil {
+		return fmt.Errorf("Unable to calculate the plan: %w", err)
+	}
+	switch {
+	case tc.PullRequest != nil:
+		if params.Name == "" {
+			// The user did not provide a name.
+			// Generate unique name base on the PR number
+			params.Name = fmt.Sprintf("pr-%d-%.8s", tc.PullRequest.Number, plan.File.ToHash)
+		}
+		planURL, err := url.JoinPath(params.Repo, "plans", params.Name)
+		if err != nil {
+			return fmt.Errorf("failed to generate plan URL: %w", err)
+		}
+		// Check if the plan already exists
+		switch _, err = a.Atlas.SchemaPlanPull(ctx, &atlasexec.SchemaPlanPullParams{
+			ConfigURL: params.ConfigURL,
+			Env:       params.Env,
+			Vars:      params.Vars,
+			URL:       planURL,
+		}); {
+		// There is a plan
+		case err == nil:
+			a.Infof("Schema plan already exists, linting the plan %q", params.Name)
+			plan, err = a.Atlas.SchemaPlanLint(ctx, &atlasexec.SchemaPlanLintParams{
+				ConfigURL: params.ConfigURL,
+				Env:       params.Env,
+				Vars:      params.Vars,
+				Context:   params.Context,
+				DevURL:    params.DevURL,
+				From:      params.From,
+				To:        params.To,
+				Repo:      params.Repo,
+				File:      planURL,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to get the schema plan: %w", err)
+			}
+			// FIXME(giautm): Remove this workaround after fixing the schema URL issue.
+			plan.File.URL = planURL
+		// The plan does not exist
+		case strings.Contains(err.Error(), fmt.Sprintf("plan %q was not found", params.Name)):
+			a.Infof("Schema plan does not exist, creating a new one with name %q", params.Name)
+			params.DryRun = false // Disable dry-run
+			params.Pending = true // Submit as pending
+			plan, err = a.Atlas.SchemaPlan(ctx, params)
+			if err != nil {
+				return fmt.Errorf("failed to save schema plan: %w", err)
+			}
+		// Unknown error
+		default:
+			return err
+		}
+		summary, err := schemaPlanComment(plan, params.Env)
+		if err != nil {
+			return fmt.Errorf("failed to generate schema plan comment: %w", err)
+		}
+		a.AddStepSummary(summary)
+		// Comment on the PR
+		err = a.GithubClient(tc.Repo, tc.SCM.APIURL).
+			UpsertComment(ctx, tc.PullRequest, plan.File.FromHash, summary)
+		if err != nil {
+			a.Errorf("failed to comment on the pull request: %v", err)
+		}
+		return nil
+	default:
```

Comment:
Do assume that the action will always run either on a PR or after merge to master?
What if someone runs the action manually on a branch?